feat: Add local proxy section, and first lab object-store-replication (#50)
flipt-io/labs: local-proxy/README.md

Flipt Local Proxy
------------

> This lab section's intent is to explore the ways that clients can achieve fast evaluations from Flipt for their feature flags. The use case here is that users would like data locality for their feature flag state, so they do not have to worry about network + application time to get a feature flag evaluation value especially for their most critical application paths.

Reviewer: ```suggestion
The intent of this lab is to explore the ways that clients can achieve fast evaluations from Flipt for their feature flags. The use case here is that users would like data locality for their feature flag state, so they do not have to worry about network latency to evaluate a feature, especially for their most critical application paths.
```

> This lab section's intent is to explore the ways that clients can achieve fast evaluations from Flipt for their feature flags. The use case here is that users would like data locality for their feature flag state, so they do not have to worry about network + application time to get a feature flag evaluation value especially for their most critical application paths.

Reviewer: looking good. I think we could potentially use some diagrams using either Monodraw or d2 like @GeorgeMac did in the cup lab: https://github.com/flipt-io/labs/blob/main/cup/argo/diagram.d2

Would really help explain whats going on I think

> ### S3 Replication

Reviewer: what if we rename this to `object store` and the folder as well? will go with how we call it in our docs: https://www.flipt.io/docs/configuration/storage#object s3 is just an implementation detail I think

This project is housed under the `s3-replication` directory. The purpose is for a user to run their application which depends on feature flags with a sidecar Flipt process that pulls data from an S3 bucket. The evaluation data here can then be accessed over `localhost` talking to the Flipt sidecar.

Lets get started!

Prerequisites:

> - [docker](https://www.docker.com/)
> - [kubectl](https://kubernetes.io/docs/reference/kubectl/)
> - [minikube](https://minikube.sigs.k8s.io/docs/)

Reviewer: ```suggestion
- [Docker](https://www.docker.com/)
- [Kubectl](https://kubernetes.io/docs/reference/kubectl/)
- [Minikube](https://minikube.sigs.k8s.io/docs/)
```

> This project uses kubernetes to deploy the following:

Reviewer: ```suggestion
This project uses Kubernetes to deploy the following:
```

- `minio`: Object store that has an S3 compatible API
- `flipt-master`: Serves as the main Flipt application, this is where users will be accessing the UI to make relevant changes
- `sample-app`: Serves as the pod with Flipt running as a sidecar, `flipt-sidecar`, pulls its data from S3 (main application can be one of your containers within the pod)
- `flipt-exporter`: CronJob that runs on a 1 minute interval that exports data out of the Flipt master, and puts those changes onto the object store

To access the `flipt-master` API, you can use kubernetes to port-forward the service:

```bash
$ kubectl port-forward svc/flipt-master --namespace default 8080:8080
```

This is so you can use the API to access the UI, and add data to Flipt as necessary.

The same goes for the sidecar:

```bash
$ kubectl port-forward svc/sample-app --namespace default 8080:8080
```

> The difference for the sidecar is that it will be in `readonly` mode, due to it sourcing its data from a file system.

> There exists a deploy script `deploy.sh` that will provision the cluster for you and start all the necessary deployments and services for the whole application described above.

Reviewer: mind renaming this script to `scripts/start` to match our existing pattern in the other labs?

<img src="./object-store-replication/diagrams/diagram.svg" alt="Object Store Replication" width="500px" />

This project is housed under the `object-store-replication` directory. The purpose is for a user to run their application which depends on feature flags with a sidecar Flipt process that pulls data from an Object Store (S3 bucket). The evaluation data here can then be accessed over `localhost` by talking to the Flipt sidecar.

Lets get started!

Prerequisites:
- [Docker](https://www.docker.com/)
- [Kubectl](https://kubernetes.io/docs/reference/kubectl/)
- [Minikube](https://minikube.sigs.k8s.io/docs/)

This project uses Kubernetes to deploy the following:
- `minio`: Object store that has an S3 compatible API
- `flipt-master`: Serves as the main Flipt application, this is where users will be accessing the UI to make relevant changes
- `sample-app`: Serves as the pod with Flipt running as a sidecar, `flipt-sidecar`. There is also a container called `evaluation-client` that will make evaluation calls to the sidecar
- `flipt-exporter`: CronJob that runs on a 1-minute interval that exports data out of the Flipt master, and puts those changes into the object store

> To access the `flipt-master` API, you can use Kubernetes to port-forward the service:

Reviewer: im wondering if we should move these instructions under the bit that talks about running `scripts/start`, as going through this, once I run `scripts/start` im not sure what im actually supposed to do next.

Maybe restructuring these as steps 1., 2., 3. etc and moving them undre the `scripts/start` would help?

Author: @markphelps Agreed. Can do that.
